src/base/matrix.rs

```diff
@@ -195,9 +195,34 @@ pub struct Matrix<T, R, C, S> {
     _phantoms: PhantomData<(T, R, C)>,
 }
 
-impl<T, R: Dim, C: Dim, S: fmt::Debug> fmt::Debug for Matrix<T, R, C, S> {
-    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
-        self.data.fmt(formatter)
+impl<T: fmt::Debug, R: Dim, C: Dim, S: RawStorage<T, R, C> + fmt::Debug> fmt::Debug
+    for Matrix<T, R, C, S>
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        if f.alternate() {
+            writeln!(f, "")?;
+        }
+        write!(f, "[")?;
+        let nrows = self.nrows();
+        let ncols = self.ncols();
+        for i in 0..nrows {
+            for j in 0..ncols {
+                if j != 0 {
+                    write!(f, ", ")?;
+                }
+                // Safety: the indices are within range
+                unsafe {
+                    (*self.data.get_unchecked(i, j)).fmt(f)?;
+                }
+            }
+            if i != nrows - 1 {
+                write!(f, "; ")?;
+            }
+            if f.alternate() && i != nrows - 1 {
+                write!(f, "\n ")?;
+            }
+        }
+        write!(f, "]")
```

Comment:
Is there a more straightforward / nicer way of looping through in this order?

Reply:
Maybe, but I wouldn't overthink it; this seems fine.

Reply:
I think it's fine: it's clear what's going on. Is the `unsafe` really necessary though? It seems to me that the bounds checks here won't lead to a very significant performance penalty given the cost of formatting etc.

Reply:
Good point, I've replaced it with `self[(i, j)].fmt(f)?;`

Reply:
I realized that the original code (with `unsafe`) is in fact unsound for 0-row matrices. It will still panic with your replacement. It's a bit tricky to get this right. Here's a quick attempt I made, which I *think* should work in all cases?

(btw, would be great to have some unit tests that also cover the 0-dimension corner cases)

```rust
        if f.alternate() {
            writeln!(f, "")?;
        }
        write!(f, "[")?;
        
        let row_separator = match f.alternate() {
            true => ";\n ",
            false => "; ",
        };
        
        for (i, row) in self.row_iter().enumerate() {
            if i > 0 {
                write!(f, "{row_separator}")?;
            }
            for (j, element) in row.iter().enumerate() {
                if j > 0 {
                    write!(f, ", ")?;
                }
                element.fmt(f)?;
            }
        }
        write!(f, "]")
```

Reply:
Oh, and also be wary of the overflow for evaluating `nrows - 1` if `nrows == 0`!

Reply:
Hmm, wouldn't the for loops only run the code if the dimensions are non-zero?

Reply:
That code looks good to me, I've added it in.

Reply:
> Hmm, wouldn't the for loops only run the code if the dimensions are non-zero? `for i in 0..0` never runs for instance. Good catch on nrows - 1

Yeah, you're right about that! Didn't think that through :-)